tests/mock/agents.py

```diff
@@ -1,221 +1,153 @@
 from unittest.mock import Mock
 
-from nucypher.blockchain.economics import EconomicsFactory
-from nucypher.blockchain.eth.agents import NucypherTokenAgent, PolicyManagerAgent, StakingEscrowAgent, WorkLockAgent
+from nucypher.blockchain.eth import agents
+from nucypher.blockchain.eth.agents import ContractAgency, EthereumContractAgent
 from nucypher.blockchain.eth.constants import NULL_ADDRESS
+from nucypher.blockchain.eth.decorators import ContractInterfaces
 from nucypher.blockchain.eth.interfaces import BlockchainInterfaceFactory
 from tests.constants import MOCK_PROVIDER_URI
 
 MOCK_TESTERCHAIN = BlockchainInterfaceFactory.get_or_create_interface(provider_uri=MOCK_PROVIDER_URI)
 CURRENT_BLOCK = MOCK_TESTERCHAIN.w3.eth.getBlock(block_identifier='latest')
 
-#
-# Fixtures
-#
-
-FAKE_RECEIPT = {'transactionHash': HexBytes(b'FAKE29890FAKE8349804'),
-                'gasUsed': 1,
-                'blockNumber': CURRENT_BLOCK.number,
-                'blockHash': HexBytes(b'FAKE43434343FAKE43443434')}
-
-
-def fake_transaction(*_a, **_kw) -> dict:
-    return FAKE_RECEIPT
 
+class MockContractAgent:
 
-def fake_call(*_a, **_kw) -> 1:
-    return 1
+    FAKE_RECEIPT = {'transactionHash': HexBytes(b'FAKE29890FAKE8349804'),
+                    'gasUsed': 1,
+                    'blockNumber': CURRENT_BLOCK.number,
+                    'blockHash': HexBytes(b'FAKE43434343FAKE43443434')}
 
+    FAKE_CALL_RESULT = 1
 
-#
-# Agents
-#
+    # Internal
+    __COLLECTION_MARKER = "contract_api"  # decorator attribute
+    __DEFAULTS = {
+        ContractInterfaces.CALL: FAKE_CALL_RESULT,
+        ContractInterfaces.TRANSACTION:  FAKE_RECEIPT
+    }
 
+    _MOCK_METHODS = list()
+    _REAL_METHODS = list()
 
-class MockContractAgent:
+    # Mock Nucypher Contract API
+    contract = Mock()
+    contract_address = NULL_ADDRESS
 
-    # Internal
+    # Mock Blockchain Interfaces
     registry = Mock()
     blockchain = MOCK_TESTERCHAIN
 
-    contract = Mock()
-    contract_address = NULL_ADDRESS
+    def __init__(self, agent_class: Type[EthereumContractAgent]):
+        """Bind mock agent attributes to the *subclass* with default values"""
+        self.agent_class = agent_class
+        self.__setup_mock(agent_class=agent_class)
+
+    @classmethod
+    def __setup_mock(cls, agent_class: Type[EthereumContractAgent]) -> None:
+        mock_methods, real_methods = list(), list(cls.__collect_real_methods(agent_class=agent_class))
+        for agent_method in real_methods:
 
-    # API
-    ATTRS = dict()
-    CALLS = tuple()
-    TRANSACTIONS = tuple()
+            # Get default effect
+            interface = getattr(agent_method, cls.__COLLECTION_MARKER)
+            default_return = cls.__DEFAULTS.get(interface)
 
-    # Spy
-    _SPY_TRANSACTIONS = defaultdict(list)
-    _SPY_CALLS = defaultdict(list)
+            # Setup Mock - Carry over the decorator marker to the mock
+            mock = Mock(return_value=default_return)
```

Comment:
I think some of them should be `PropertyMock`

Reply:
Let me take a closer look. Two thoughts here:

1. We are overriding the name namespce of the property with a `Mock` explicitly so I am surprised we need a PropertyMock at all.

2.  We need a clear convention for when to use a property to expose a contract interface: Perhaps we use properties for *all contract attributes?*

Reply:
Turns out be nontrival.  I pushed an update that I think will work without an explicit `PropertyMock`.